firebase-firestore/src/main/java/com/google/firebase/firestore/core/UserData.java

```diff
@@ -303,9 +301,17 @@ private void validatePath() {
       }
     }
 
+    private static final String RESERVED = "__";
+
     private void validatePathSegment(String segment) {
-      if (isWrite() && reservedFieldRegex.matcher(segment).find()) {
-        throw this.createError("Document fields cannot begin and end with __");
+      if (!isWrite()) return;
```

Comment:
I think we may want to prevent empty field names even for non-writes...?  E.g. `.whereEqualTo("", "value")` should probably fail.  But we don't want to validate the __.*__ regex since `.whereEqualTo("__key__", "x/y/z")` is valid.